rust/bridge/jni/src/logging.rs

```diff
@@ -0,0 +1,145 @@
+impl From<JavaLogLevel> for log::Level {
+    fn from(level: JavaLogLevel) -> Self {
+        use JavaLogLevel::*;
+        match level {
+            Error | Assert => Self::Error,
+            Warn => Self::Warn,
+            Info => Self::Info,
+            Debug => Self::Debug,
+            Verbose => Self::Trace,
+        }
+    }
+}
+
+struct JniLogger {
+    vm: JavaVM,
+    logger_class: GlobalRef,
+}
+
+impl JniLogger {
+    fn log_impl(&self, record: &log::Record) -> jni::errors::Result<()> {
+        let env = self.vm.attach_current_thread()?;
+        let level: JavaLogLevel = record.level().into();
+        let message = format!(
+            "{}:{}: {}",
+            record.file().unwrap_or("<unknown>"),
+            record.line().unwrap_or(0),
+            record.args(),
+        );
+        let args: [JValue; 3] = [
+            level.into(),
+            env.new_string("libsignal-client")?.into(),
+            env.new_string(message)?.into(),
+        ];
+        let result = env.call_static_method(
+            &self.logger_class,
+            "log",
+            "(ILjava/lang/String;Ljava/lang/String;)V",
+            &args,
+        );
+
+        let throwable = env.exception_occurred()?;
+        if **throwable == *JObject::null() {
+            result?;
+        } else {
+            env.exception_clear()?;
+        }
+        Ok(())
+    }
+}
+
+impl log::Log for JniLogger {
+    fn enabled(&self, _metadata: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        if self.log_impl(record).is_err() {
+            // Drop the error; it's not like we can log it!
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_org_signal_client_internal_Native_Logger_1Initialize(
+    env: JNIEnv,
+    _class: JClass,
+    max_level: jint,
+    logger_class: JClass,
+) {
+    let logger = JniLogger {
+        vm: env.get_java_vm().expect("could not initialize logging"),
+        logger_class: env
+            .new_global_ref(logger_class)
+            .expect("could not initialize logging"),
+    };
+
+    // Keep this in sync with SignalProtocolLogger.java.
+    let level = match max_level {
+        // The jni crate uses debug! in its own implementation.
+        2 | 3 => panic!("invalid log level (must be INFO or higher for libsignal-client)"),
+        4 => JavaLogLevel::Info,
+        5 => JavaLogLevel::Warn,
+        6 => JavaLogLevel::Error,
+        7 => JavaLogLevel::Assert,
+        _ => panic!("invalid log level (see SignalProtocolLogger)"),
+    };
+    assert!(jint::from(level) == max_level);
+
+    match log::set_logger(Box::leak(Box::new(logger))) {
+        Ok(_) => {
+            log::set_max_level(log::Level::from(level).to_level_filter());
+            log::debug!("logging initialized for libsignal-client");
+        }
+        Err(_) => {
+            log::warn!("logging already initialized for libsignal-client; ignoring later call");
```

Comment:
Can we here still call `set_max_level`? That allows Android to use `Logger_Initialize(SignalProtocolLogger.DEBUG);` to enable debug logging if they need it. This unavoidably leaks the new, unused, logger implementation but presumably Android would only ever call this one extra time.

Alternately, split it into two functions?

`Logger_Initialize(Log.class)` // called just once, in `static` block
`Logger_SetMaxLevel(SingalProtocolLogger.DEBUG)` // called with `INFO` in static block, potentially called again later by Android if they need it

Then there is no leak and Android can switch debug/info logging however often they like.

Reply:
The default max level is "off", so I think it's correct to have it as an argument here, but I can expose a second entry point for it as well, sure.